ethexe/signer/src/digest.rs

```diff
@@ -0,0 +1,255 @@
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+//! Keccak256 digest type. Implements AsDigest hashing for ethexe common types.
+
+use core::fmt;
+use ethexe_common::router::{
+    BlockCommitment, CodeCommitment, OutgoingMessage, StateTransition, ValueClaim,
+};
+use gprimitives::{MessageId, H256};
+use parity_scale_codec::{Decode, Encode};
+use sha3::Digest as _;
+
+use crate::Address;
+
+#[derive(
+    Clone,
+    Copy,
+    PartialOrd,
+    Ord,
+    PartialEq,
+    Eq,
+    Hash,
+    Encode,
+    Decode,
+    derive_more::From,
+    derive_more::Into,
+    derive_more::AsRef,
+)]
+pub struct Digest([u8; 32]);
+
+impl fmt::Debug for Digest {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "0x{}", hex::encode(self.0))
+    }
+}
+
+impl fmt::Display for Digest {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "0x{}", hex::encode(self.0))
+    }
+}
+
+/// Trait for hashing types into a Digest using Keccak256.
+pub trait AsDigest {
+    fn as_digest(&self) -> Digest;
+}
+
+impl AsDigest for Digest {
+    fn as_digest(&self) -> Digest {
+        *self
+    }
+}
+
+impl<T: AsDigest> AsDigest for [T] {
+    fn as_digest(&self) -> Digest {
+        let mut message = Vec::with_capacity(self.len() * size_of::<Digest>());
+
+        for item in self.iter() {
+            message.extend_from_slice(item.as_digest().as_ref());
+        }
+
+        message.as_digest()
+    }
+}
+
+impl<T: AsDigest> AsDigest for Vec<T> {
+    fn as_digest(&self) -> Digest {
+        self.as_slice().as_digest()
+    }
+}
+
+impl AsDigest for [u8] {
+    fn as_digest(&self) -> Digest {
+        Digest(sha3::Keccak256::digest(self).into())
+    }
+}
+
+impl AsDigest for CodeCommitment {
+    fn as_digest(&self) -> Digest {
+        self.encode().as_digest()
+    }
+}
+
+impl AsDigest for StateTransition {
+    fn as_digest(&self) -> Digest {
+        // State transition basic fields.
+
+        let state_transition_size = // concat of fields:
+            // actorId
+            size_of::<Address>()
+            // newStateHash
+            + size_of::<H256>()
+            // valueToReceive
+            + size_of::<u128>()
+            // digest(valueClaimsBytes)
+            + size_of::<Digest>()
+            // digest(messagesHashesBytes)
+            + size_of::<Digest>();
+
+        let mut state_transition_bytes = Vec::with_capacity(state_transition_size);
+
+        state_transition_bytes.extend_from_slice(self.actor_id.to_address_lossy().as_bytes());
+        state_transition_bytes.extend_from_slice(self.new_state_hash.as_bytes());
+        state_transition_bytes.extend_from_slice(self.value_to_receive.to_be_bytes().as_slice());
```

Comment:
```suggestion
        let state_transition_size = [
        	self.actor_id.to_address_lossy().as_bytes(),
		self.new_state_hash.as_bytes(),
		self.value_to_receive.to_be_bytes().as_slice(),
	]
	.concat()
```
`concat()` does the same

Reply:
We pre-allocate bytes for all here and `concat` in your suggestions will preallocate only for actor_id new_state_hash and value_to_receive

Reply:
Create the rest before `concat` then

Reply:
Done